### DISCUSSION TITLE
Discussion: Withers and collections

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#3262] [Language Ideas]
As a pre-requirement, we need to relax with-expression syntax to accept nested initializers:

 
```cs
record with { Owner = { Name = "name" } }
// -> 
record.With(owner: record.Owner.With(name: "name"))
```

This is using the exact same grammar as object/collection initializers. 
For an extended form of the syntax, see #77.

Then we can accept a collection initializer with the following generated code:

```cs
node with { ArgList = { Args = { e } } }
// -> 
node.With(argList: node.ArgList.With(args: node.ArgList.Args.Add(e))
```

Such code is already accepted by object/collection initializer syntax so we can use the same parsing rules.

Currently the collection initializer doesn't care about the return type on `Add` method. In withers, we require the `Add` method to return a new object to be passed to With method.

To avoid a new allocation on every addition, we could say if there is an `AddRange` method, we'll use that instead,

```cs
node with { ArgList = { Args = { e, a } } }
// ->
node.With(argList: node.ArgList.With(args: node.ArgList.Args.AddRange(e, a))
```

Note: if we wanted a replacement, we can write
```cs
node with { ArgList = { Args = new() { ... } } }
```
That's exactly how collection initializers work today.

**Comment by agocke** [comment]
An alternate syntax, which may prevent syntactic ambiguity, could be

```C#
record with { Owner with { Name = "name" } }
```

**Comment by alrz** [comment]
@agocke 

I don't think there is any? I'm specifically thinking `InitializerExpressionSyntax`, not any new syntax - covers both collection and regular usage.

**Comment by agocke** [comment]
Maybe there isn't one, just recording some options in case.

**Comment by orthoxerox** [comment]
See also #77

**Comment by alrz** [comment]
@orthoxerox forgot to mention that issue. This is supposed to be a tiny increment to that proposal. Updated.

**Comment by qrli** [comment]
> ```cs
> record with { Owner with { Name = "name" } }
> ```

As `with` implies a new object is created, I read it as another `Owner` object is created and assigned to `Owner` property of newly created record.

**Comment by svick** [comment]
I don't like that the syntax `Args = { e }` is used to **add** an item to an existing collection, when it looks like it should be setting the items in the collection.

Collection initializers can get away with that, because they're usually adding items to an empty collection (or at least some kind of "default" collection), but that's not the case here.

**Comment by alrz** [comment]
If that's confusing, I don't think `Args with { e }` would be much an improvement, to me, it's not still clear that we're adding items and not replacing.

I think if you want to be completely clear, there should be a way to specify the exact operation, currently, the one time you need to `Remove` an item, you should fallback to the usual syntax

```cs
node with { ArgList = { Args = node.ArgList.Args.Remove(itemOrIndex) } }
```
Maybe `{ ArgList with { Args = .Remove(itemOrIndex) } }` but it doesn't quite look any better?

**Comment by qrli** [comment]
I can accept the "add" behavior because wither does creation, so is collection initializer. 
It is true that the collection may not be empty in wither, but so does collection initializer, even though it is a bit rare.

I agree the `Args = { e }` syntax is a bit confusing, but for collection initializer too. It is better to keep consistent. If it were to be improved, collection initializer is better to have the same improvement.
